pkg/memory/short_term.go

```diff
@@ -0,0 +1,150 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains the implementation of the memory where the healer stores the list of actions
+// that have already been performed.
+
+package memory
+
+import (
+	"reflect"
+	"sync"
+	"time"
+)
+
+// ShortTermMemoryBuilder builds of short term memory objects.
+//
+type ShortTermMemoryBuilder struct {
+	// How long to remember actions.
+	duration time.Duration
+}
+
+// ShortTermMemory stores a set of items for a given period of time.
+//
+type ShortTermMemory struct {
+	// How long to remember actions.
+	duration time.Duration
+
+	// There will be a cell for each action stored, containing the action itself and the time it was
+	// added to the memory.
+	cells []*ShortTermCell
+
+	// Mutex used to prevent simultaneous updates of the data structures.
+	mutex *sync.Mutex
+}
+
+// ShortTermCell stores each individual action, and the time it was added to the memory.
+//
+type ShortTermCell struct {
+	// The item stored in the cell.
+	item interface{}
+
+	// The time when the cell was created or updated.
+	stamp time.Time
+}
+
+// NewShortTermMemoryBuilder creates a builder that can create short term memory objects.
+//
+func NewShortTermMemoryBuilder() *ShortTermMemoryBuilder {
+	b := new(ShortTermMemoryBuilder)
+	return b
+}
+
+// Duration sets how long objects in the memory will be remembered. The default is zero, which means
+// that objects won't be remembered at all.
+//
+func (b *ShortTermMemoryBuilder) Duration(duration time.Duration) *ShortTermMemoryBuilder {
+	b.duration = duration
+	return b
+}
+
+// Build creates a new short term memory object with the configuration stored in the builder.
+//
+func (b *ShortTermMemoryBuilder) Build() (m *ShortTermMemory, err error) {
+	m = new(ShortTermMemory)
+	m.duration = b.duration
+	m.cells = make([]*ShortTermCell, 0)
+	m.mutex = &sync.Mutex{}
+	return
+}
+
+// Add adds a new item to the memory.
+//
+func (m *ShortTermMemory) Add(item interface{}) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	cell := m.findMatchingCell(item)
+	if cell == nil {
+		cell = new(ShortTermCell)
+		cell.item = item
+		m.cells = append(m.cells, cell)
+	}
+	cell.stamp = time.Now()
+}
+
+func (m *ShortTermMemory) Has(item interface{}) bool {
+	// Purge cells before checking.
+	m.PurgeExpiredCells()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.findMatchingCell(item) != nil
+}
+
+// Len returns the number of items inside the memory.
+//
+func (m *ShortTermMemory) Len() int {
+	m.PurgeExpiredCells()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return len(m.cells)
+}
+
+// PurgeExpiredCells finds the aged cells and removes them.
+//
+func (m *ShortTermMemory) PurgeExpiredCells() {
```

Comment:
suggestion: make this private lock-free, and wrap it with a public method that acquires the lock.
Then, from `Len()`, `Has()` and other methods, call the lock-free private `purge` from inside the lock context.
This will save lock acquiring time

Reply:
I kind of like it this way :( 
@jhernand  what do you think?

Reply:
I guess that you made this method public in order to be able to call it from the main program. I think that isn't ideal, as the details of purging the memory should better be hidden behind the memory object. I'd suggest to make it private. The suggestion to make it lock-free also makes sense to me.